### PR TITLE
Support empty projection in CSV and JSON readers

### LINKED ISSUE #2603: Support empty projection in CSV, JSON readers
**Is your feature request related to a problem or challenge? Please describe what you are trying to do.**
As a user of the readers, I want to be able to use the CSV and JSON readers with empty projection.

**Describe the solution you'd like**

Support empty projections in the readers. The readers should generate a RecordBatch that returns a batch with the correct number of rows.

**Describe alternatives you've considered**

**Additional context**
In DataFusion I want to avoid scanning a column https://github.com/apache/arrow-datafusion/issues/3214 . Implementing the support in the arrow readers seems the best and easiest way forward.

### PR DESCRIPTION
**Pull request by Dandandan**
# Which issue does this PR close?

Closes #2603

# Rationale for this change

# What changes are included in this PR?
Support empty projection

# Are there any user-facing changes?

### CODE REVIEW

**Comment by Dandandan** [comment]
Some test failures related to the changes, will have a look later

File: arrow/src/json/reader.rs [hunk 1]
```diff
@@ -1178,11 +1190,16 @@ impl Decoder {
         &self,
         rows: &[Value],
         struct_fields: &[Field],
-        projection: &[String],
+        projection: &Option<Vec<String>>,
```
Comment (HaoYang670) [inline arrow/src/json/reader.rs:1193]: BTW, we should also update the doc:
```
/// If `projection` is not empty, then all values are returned. ...
```

File: arrow/src/json/reader.rs [hunk 2]
```diff
@@ -1178,11 +1190,16 @@ impl Decoder {
...
     ) -> Result<Vec<ArrayRef>> {
         let arrays: Result<Vec<ArrayRef>> = struct_fields
             .iter()
-            .filter(|field| projection.is_empty() || projection.contains(field.name()))
+            .filter(|field| {
+                projection
+                    .as_ref()
+                    .map(|p| p.contains(field.name()))
+                    .unwrap_or(true)
```
Comment (HaoYang670) [inline arrow/src/json/reader.rs:1201]: 👍

**Review by HaoYang670 — APPROVED** [review]
LGTM.

**Review by viirya — APPROVED** [review]

**Pull request merged**